### PR TITLE
Fixes and improvements to HWM tracker + UTs.

### PR DESCRIPTION
**Pull request by fasaxc**
- Add missing DeleteSubtree() call to HWM tracker.
- Handle tracking of nested deletions.
- Add UTs.

### CODE REVIEW

File: lib/hwm/hwm.go [hunk 1]
```diff
@@ -120,6 +139,15 @@ func (trie *HighWatermarkTracker) DeleteOldKeys(hwmLimit uint64) []string {
 	return deletedKeys
 }
 
+func (trie *HighWatermarkTracker) ToMap() map[string]uint64 {
```
Comment (matthewdupre) [inline lib/hwm/hwm.go]: It's sad that this function is only used by the tests.  Does it have to be here?

Comment (fasaxc) [inline lib/hwm/hwm.go]: Alternative is to put the tests in the same package and poke into private state.  I think this is a bit cleaner.

Comment (caseydavenport) [inline lib/hwm/hwm.go]: Hm, I thought the Go test philosophy was to include tests in the same package as the thing they tested anyway (though I could be wrong!)

Comment (fasaxc) [inline lib/hwm/hwm.go]: There are two types of tests: you can put tests in the same package `foo`, say or in a special test package in the same directory `foo_test`.  The latter type test using your public API, so more of a black box test.  Both sorts of test are favoured.  In this case, it's borderline but adding a new (perfectly reasonable) public method seemed a lot cleaner than poking into the awkward internals.

**Comment by robbrockbank** [comment]
lgtm - let's get this merged.

**Pull request merged**